src/main/java/frc/trigon/robot/poseestimation/apriltagcamera/AprilTagCamera.java

```diff
@@ -0,0 +1,186 @@
+        return tagPose.getTranslation().toTranslation2d().plus(fieldRelativeTagPose);
+    }
+
+    private Translation2d getTagToCamera(Pose3d tagPose) {
+        final double cameraToTagDistanceMeters = -PhotonUtils.calculateDistanceToTargetMeters(
+                robotCenterToCamera.getZ(), tagPose.getZ(), robotCenterToCamera.getRotation().getY(), inputs.bestTargetRelativePitchRadians
+        );
+        final double cameraToTagXDistance = Math.sin(inputs.bestTargetRelativeYawRadians + robotCenterToCamera.getRotation().getZ()) * cameraToTagDistanceMeters;
+        final double cameraToTagYDistance = -Math.cos(inputs.bestTargetRelativeYawRadians + robotCenterToCamera.getRotation().getZ()) * cameraToTagDistanceMeters;
+        return new Translation2d(cameraToTagXDistance, cameraToTagYDistance);
+    }
+
+    private Pose2d cameraPoseToRobotPose(Pose3d cameraPose) {
+        if (cameraPose == null)
+            return null;
+        return cameraPose.transformBy(robotCenterToCamera.inverse()).toPose2d();
+    }
+
+    private boolean isNewTimestamp() {
+        if (lastUpdatedTimestamp == getLastResultTimestamp())
+            return false;
+
+        lastUpdatedTimestamp = getLastResultTimestamp();
+        return true;
+    }
+
+    private void logVisibleTags() {
+        if (inputs.hasResult) {
+            Logger.recordOutput("VisibleTags/" + this.getName(), new Pose2d[0]);
+            return;
+        }
+
+        final Pose2d[] visibleTagPoses = new Pose2d[inputs.visibleTagIDs.length];
+        for (int i = 0; i < visibleTagPoses.length; i++)
+            visibleTagPoses[i] = FieldConstants.TAG_ID_TO_POSE.get(i).toPose2d();
+        Logger.recordOutput("VisibleTags/" + this.getName(), visibleTagPoses);
+    }
+
+    private Matrix<N3, N1> calculateSolvePNPStdDevs() {
+        final int numberOfVisibleTags = inputs.visibleTagIDs.length;
+        final double translationStd = calculateStd(solvePNPTranslationsStdExponent, inputs.averageDistanceFromAllTags, numberOfVisibleTags);
+        final double thetaStd = calculateStd(solvePNPThetaStdExponent, inputs.averageDistanceFromAllTags, numberOfVisibleTags);
+
+        return VecBuilder.fill(translationStd, translationStd, thetaStd);
+    }
+
+    /**
+     * Calculates the standard deviations of the pose, for when the pose is calculated by assuming the robot's heading.
+     * The theta deviation is infinite so that it won't change anything in the heading because we assume it is correct.
+     *
+     * @return the standard deviations
+     */
+    private Matrix<N3, N1> calculateAssumedHeadingStdDevs() {
+        final double translationStd = calculateStd(assumedRobotPoseTranslationsStdExponent, inputs.distanceFromBestTag, inputs.visibleTagIDs.length);
+        final double thetaStd = Double.POSITIVE_INFINITY;
+
+        return VecBuilder.fill(translationStd, translationStd, thetaStd);
+    }
+
+    private double calculateStd(double exponent, double distance, int visibleTags) {
```

Comment:
calculateStdDev (StdDev stands for Standard Deviations, maybe we should call everything that and not std? Both approaches are valid)

Add a javadoc and explain why this works as it is, how it works, why it needs the distance and visjble tags, etc.